internal/common/base_client.go

```diff
@@ -0,0 +1,43 @@
+package common
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+)
+
+const id = 420
+
+func NewClient() Client {
+	return &Base_client{}
+}
+
+type Base_client struct {
+	id shared.ClientID
+}
+
+func (c *Base_client) Echo(s string) string {
+	c.logf("Echo: '%v'", s)
+	return s
+}
+
+func (c *Base_client) GetID() shared.ClientID {
+	return c.id
+}
+
+// logf is the client's logger that prepends logs with your ID. This makes
+// it easier to read logs. DO NOT use other loggers that will mess logs up!
+func (c *Base_client) logf(format string, a ...interface{}) {
+	log.Printf("[%v]: %v", c.id, fmt.Sprintf(format, a...))
+}
+
+func (c *Base_client) StartOfTurnUpdate(gameState GameState) {
+	c.logf("Received game state update: %v", gameState)
```

Comment:
it'd be nice to do `c.logf("BASE CLIENT IMPLEMENTATION, PLEASE OVERRIDE THIS")` since we only know at runtime that we overrode something (that needs overriding) correctly.

Reply:
I am thinking maybe using: BASE, COMPULSORY, OPTIONAL in comments above each function?

Reply:
Where BASE is dont touch, Compulsory needs to be implemented and optional is well, optional

Reply:
lgtm!